### PR TITLE
Week4

### PR DESCRIPTION
**Pull request by LouisaStrunk02**
changes:
1. add volume
2. add filter-less
3. add filter-more
4. add recover

### CODE REVIEW

File: week4/filter-less/filter.c [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    // Iterate over infile's scanlines
+    for (int i = 0; i < height; i++)
+    {
+        // Read row into pixel array
+        fread(image[i], sizeof(RGBTRIPLE), width, inptr);
+
+        // Skip over padding
+        fseek(inptr, padding, SEEK_CUR);
+    }
+
```
Comment (stefanluth) [inline week4/filter-less/filter.c:99]: `inptr` is not used afterwards, you might want to close the file here.

File: week4/filter-less/filter.c [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    // Open input file
+    FILE *inptr = fopen(infile, "r");
+    if (inptr == NULL)
+    {
+        printf("Could not open %s.\n", infile);
+        return 4;
+    }
+
+    // Open output file
+    FILE *outptr = fopen(outfile, "w");
```
Comment (stefanluth) [inline week4/filter-less/filter.c:47]: `outptr` is not needed until you actually write to the file further down (line  125+). You might want to move this to right before writing happens. That way you don't need to close the file in every if-check.

File: week4/filter-less/filter.dSYM/Contents/Info.plist [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (stefanluth) [inline week4/filter-less/filter.dSYM/Contents/Info.plist:1]: Is this file needed?

File: week4/recover/recover.c [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    int image_counter = 0;
+
+    FILE *output_file = NULL;
+
+    char *filename = malloc(8 * sizeof(char));
+
+    while (fread(buffer, sizeof(char), 512, input_file))
+    {
+        bool isStartOfJPEG = (buffer[0] == 0xff && buffer[1] == 0xd8 && buffer[2] == 0xff && (buffer[3] & 0xf0) == 0xe0);
```
Comment (stefanluth) [inline week4/recover/recover.c:35]: Keep variables consistent. (snake_case)

**Review by stefanluth — CHANGES_REQUESTED** [review]
My review comments in `filter-less/filter.c` also apply to the other files.

**Review by stefanluth — CHANGES_REQUESTED** [review]
Resolved unresolved requested changes without comments.

File: week4/filter-less/helpers.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            float pixelcounter = 0.00;
+
+            for (int x = -1; x < 2; x++)
+            {
+                for (int y = -1; y < 2; y++)
+                {
+                    int currentX = i + x;
+                    int currentY = j + y;
+
+                    bool pixel_Nonexistent = currentX < 0 || currentX >= height || currentY < 0 || currentY >= width;
```
Comment (stefanluth) [inline week4/filter-less/helpers.c:99]:
```suggestion
                    bool pixel_nonexistent = currentX < 0 || currentX >= height || currentY < 0 || currentY >= width;
```

File: week4/recover/recover.c [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    int image_counter = 0;
+
+    FILE *output_file = NULL;
+
+    char *filename = malloc(8 * sizeof(char));
+
+    while (fread(buffer, sizeof(char), 512, input_file))
+    {
+        bool is_Start_Of_JPEG = (buffer[0] == 0xff && buffer[1] == 0xd8 && buffer[2] == 0xff && (buffer[3] & 0xf0) == 0xe0);
```
Comment (stefanluth) [inline week4/recover/recover.c:35]:
```suggestion
        bool is_start_of_jpeg = (buffer[0] == 0xff && buffer[1] == 0xd8 && buffer[2] == 0xff && (buffer[3] & 0xf0) == 0xe0);
```

File: week4/filter-less/helpers.c (hunk 5 above)
Comment (stefanluth) [reply week4/filter-less/helpers.c:99]: [snake_case](https://en.wikipedia.org/wiki/Snake_case) starts each word with lowercase letters

File: week4/filter-less/filter.c (hunk 2 above)
Comment (stefanluth) [reply week4/filter-less/filter.c:47]: Resolved as file was given by task and not edited.

File: week4/filter-less/helpers.c (hunk 5 above)
Comment (stefanluth) [reply week4/filter-less/helpers.c:99]: Nvm, you use [camelCase](https://en.wikipedia.org/wiki/Camel_case) everywhere else, so use that.

Comment (stefanluth) [reply week4/filter-less/helpers.c:99]:
```suggestion
                    bool pixelNonExistent = currentX < 0 || currentX >= height || currentY < 0 || currentY >= width;
```

File: week4/filter-more/helpers.c [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+            for (int x = -1; x < 2; x++)
+            {
+                for (int y = -1; y < 2; y++)
+                {
+                    int currentX = i + x;
+                    int currentY = j + y;
+
+                    bool pixel_Nonexistent = currentX < 0 || currentX >= height || currentY < 0 || currentY >= width;
+                    if (pixel_Nonexistent)
```
Comment (stefanluth) [inline week4/filter-more/helpers.c:67]:
```suggestion
                    bool pixelNonExistent = ...
```

Comment (stefanluth) [reply week4/filter-more/helpers.c:67]: Sorry, use [camelCase](https://en.wikipedia.org/wiki/Camel_case) as done above and elsewhere.

File: week4/filter-more/helpers.c [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+    for (int i = 0; i < height; i++)
+    {
+        for (int j = 0; j < width; j++)
+        {
+            int red, green, blue;
+            red = 0;
+            green = 0;
+            blue = 0;
+            float pixelcounter = 0.00;
```
Comment (stefanluth) [inline week4/filter-more/helpers.c:57]:
```suggestion
            float pixelCounter = 0.00;
```

File: week4/filter-less/filter.c (hunk 1 above)
Comment (LouisaStrunk02) [reply week4/filter-less/filter.c:99]: In filter-more and filter-less I only had to create helpers.c myself. The other files were given by the task.

File: week4/filter-less/filter.c (hunk 2 above)
Comment (LouisaStrunk02) [reply week4/filter-less/filter.c:47]: same as above

File: week4/filter-less/filter.dSYM/Contents/Info.plist (hunk 3 above)
Comment (LouisaStrunk02) [reply week4/filter-less/filter.dSYM/Contents/Info.plist:1]: same as above

File: week4/filter-more/helpers.c [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+                    redX += image[i - 1 + x][j - 1 + y].rgbtRed * Gx[x][y];
+                    greenX += image[i - 1 + x][j - 1 + y].rgbtGreen * Gx[x][y];
+                    blueX += image[i - 1 + x][j - 1 + y].rgbtBlue * Gx[x][y];
+
+                    redY += image[i - 1 + x][j - 1 + y].rgbtRed * Gy[x][y];
+                    greenY += image[i - 1 + x][j - 1 + y].rgbtGreen * Gy[x][y];
+                    blueY += image[i - 1 + x][j - 1 + y].rgbtBlue * Gy[x][y];
+                }
+            }
+            int red = round(sqrt((redX * redX) + (redY * redY)));
```
Comment (stefanluth) [inline week4/filter-more/helpers.c:138]:
```suggestion
            }

            int red = round(sqrt((redX * redX) + (redY * redY)));
```

File: week4/volume/volume.c [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+    int16_t buffer;
+    while (fread(&buffer, sizeof(int16_t), 1, input))
+    {
+        buffer *= factor;
+        fwrite(&buffer, sizeof(int16_t), 1, output);
+    }
+
+    fclose(input);
+    fclose(output);
```
Comment (stefanluth) [inline week4/volume/volume.c:43]: Missing return.

**Review by stefanluth — CHANGES_REQUESTED** [review]
Could you add a blank line before each function's last return?

File: week4/filter-less/helpers.c [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            float green = image[i][j].rgbtGreen;
+            float blue = image[i][j].rgbtBlue;
+
+            int average = round((red + green + blue) / 3.0);
+            image[i][j].rgbtRed = average;
+            image[i][j].rgbtGreen = average;
+            image[i][j].rgbtBlue = average;
+        }
+    }
+    return;
```
Comment (stefanluth) [inline week4/filter-less/helpers.c:21]: Blank line here
```suggestion
    }

    return;
```

File: week4/filter-less/helpers.c [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+            {
+                sepiaBlue = 255;
+            }
+
+            image[i][j].rgbtRed = sepiaRed;
+            image[i][j].rgbtGreen = sepiaGreen;
+            image[i][j].rgbtBlue = sepiaBlue;
+        }
+    }
+    return;
```
Comment (stefanluth) [inline week4/filter-less/helpers.c:54]:
```suggestion
    }

    return;
```

File: week4/filter-less/helpers.c [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    for (int i = 0; i < height; i++)
+    {
+        for (int j = 0; j < width / 2; j++)
+        {
+            RGBTRIPLE temp = image[i][j];
+            image[i][j] = image[i][width - (j + 1)];
+            image[i][width - (j + 1)] = temp;
+        }
+    }
+    return;
```
Comment (stefanluth) [inline week4/filter-less/helpers.c:68]:
```suggestion
    }

    return;
```

**Review by stefanluth — COMMENTED** [review]
Examples

File: week4/volume/volume.c (hunk 10 above)
Comment (LouisaStrunk02) [reply week4/volume/volume.c:43]: A return statement results in an error message.

**Review by stefanluth — APPROVED** [review]

**Review by shoreal — CHANGES_REQUESTED** [review]
Delete the .DS_Store files. Code LGTM

**Review by DevBastiS — APPROVED** [review]
Code LGTM, do what @shoreal said, after that it is fine.

**Comment by stefanluth** [comment]
> Delete the .DS_Store files. Code LGTM

Add a `.gitignore` [file](https://www.atlassian.com/de/git/tutorials/saving-changes/gitignore) and include `.DS_Store` 👍

File: .gitignore [hunk 14]
```diff
@@ -0,0 +1 @@
+**/.DS_Store
```
Comment (shoreal) [inline .gitignore:1]: EOF missing 🙈😅

**Review by shoreal — CHANGES_REQUESTED** [review]

**Pull request merged**